text/0000-simd-infrastructure.md

```diff
@@ -0,0 +1,422 @@
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Lay the ground work for building powerful SIMD functionality.
+
+# Motivation
+
+SIMD (Single-Instruction Multiple-Data) is an important part of
+performant modern applications. Most CPUs used for that sort of task
+provide dedicated hardware and instructions for operating on multiple
+values in a single instruction, and exposing this is an important part
+of being a low-level language.
+
+This RFC lays the ground-work for building nice SIMD functionality,
+but doesn't fill everything out. The goal here is to provide the raw
+types and access to the raw instructions on each platform.
+
+(An earlier variant of this RFC was discussed as a
+[pre-RFC](https://internals.rust-lang.org/t/pre-rfc-simd-groundwork/2343).)
+
+## Where does this code go? Aka. why not in `std`?
+
+This RFC is focused on building stable, powerful SIMD functionality in
+external crates, not `std`.
+
+This makes it much easier to support functionality only "occasionally"
+available with Rust's preexisting `cfg` system. There's no way for
+`std` to conditionally provide an API based on the target features
+used for the final artifact. Building `std` in every configuration is
+certainly untenable. Hence, if it were to be in `std`, there would
+need to be some highly delayed `cfg` system to support that sort of
+conditional API exposure.
+
+With an external crate, we can leverage `cargo`'s existing build
+infrastructure: compiling with some target features will rebuild with
+those features enabled.
+
+
+# Detailed design
+
+The design comes in three parts, all on the path to stabilisation:
+
+- types (`feature(repr_simd)`)
+- operations (`feature(platform_intrinsics)`)
+- platform detection (`feature(cfg_target_feature)`)
+
+The general idea is to avoid bad performance cliffs, so that an
+intrinsic call in Rust maps to preferably one CPU instruction, or, if
+not, the "optimal" sequence required to do the given operation
+anyway. This means exposing a *lot* of platform specific details,
+since platforms behave very differently: both across architecture
+families (x86, x86-64, ARM, MIPS, ...), and even within a family
+(x86-64's Skylake, Haswell, Nehalem, ...).
+
+There is definitely a common core of SIMD functionality shared across
+many platforms, but this RFC doesn't try to extract that, it is just
+building tools that can be wrapped into a more uniform API later.
+
+
+## Types
+
+There is a new attribute: `repr(simd)`.
+
+```rust
+#[repr(simd)]
+struct f32x4(f32, f32, f32, f32);
+
+#[repr(simd)]
+struct Simd2<T>(T, T);
+```
+
+The `simd` `repr` can be attached to a struct and will cause such a
+struct to be compiled to a SIMD vector. It can be generic, but it is
+required that any fully monomorphised instance of the type consist of
+only a single "primitive" type, repeated some number of times. Types
+are flattened, so, for `struct Bar(u64);`, `Simd2<Bar>` has the same
+representation as `Simd2<u64>`.
```

Comment:
Hm, are you visualising something like the following being valid?

```rust
struct Foo(f32);
#[repr(simd, flatten)]
struct Foox4(Foo, Foo, Foo, Foo);
```

(And if there was no `flatten`, it would be invalid?)

---

In any case, I'm actually going to remove this aspect of the RFC: I'm not sure it's pulling its weight: it is moderately non-trivial for me to implement, and my experimentations have ended up working well without it.